### PR TITLE
[Core] Remove lora additional vocabulary

### PR DESCRIPTION
**Pull request by ahengljh**
## Purpose

  This PR removes the ability for each LoRA adapter to define its own additional
  vocabulary, as proposed in issue #23474. All LoRA adapters for a given model now
  share the same vocabulary as the base model.

  ## Test Plan

  1. Test basic LoRA initialization:
  ```python
  import vllm
  llm = vllm.LLM(model="Qwen3-8B", enable_lora=True)
```

  2. Test LoRA adapter loading (with existing adapters):
  ```python
  from vllm.lora.request import LoRARequest
  lora_request = LoRARequest("adapter1", 1, "/path/to/lora/adapter")
  outputs = llm.generate(["Test prompt"], lora_request=lora_request)
```

  4. Run LoRA-related tests:
  ```python
  pytest tests/lora/test_lora_manager.py -v
  pytest tests/lora/test_layers.py -v 
```

  ## Test Result

  - ✅ Successfully initialized vLLM with enable_lora=True without AttributeError
  - ✅ LoRAConfig no longer has lora_extra_vocab_size attribute
  - ✅ All model files updated to remove additional vocabulary references

  (Optional) Documentation Update

  No documentation updates required as this removes an undocumented internal feature.
  <details>
  <summary> Essential Elements of an Effective PR Description Checklist </summary>
  - [X] The purpose of the PR, such as "Fix some issue (link existing issues this PR
  will resolve)".
  - [X] The test plan, such as providing test command.
  - [X] The test results, such as pasting the results comparison before and after, or
  e2e results
  - [ ] (Optional) The necessary documentation update, such as updating
  `supported_models.md` and `examples` for a new model.
 </details>
  BEFORE SUBMITTING, PLEASE READ https://docs.vllm.ai/en/latest/contributing

### CODE REVIEW

**Comment by ahengljh** [comment]
Allow me to fix some naive mistake here

**Comment by jeejeelee** [comment]
As a start, maybe it would be better to just provide a deprecation warning for `lora_extra_vocab_size`

**Comment by ahengljh** [comment]
> As a start, maybe it would be better to just provide a deprecation warning for `lora_extra_vocab_size`

Oh I see. So we should hold this PR and just add warning info to users at this moment right?

**Comment by jeejeelee** [comment]
> > As a start, maybe it would be better to just provide a deprecation warning for `lora_extra_vocab_size`
> 
> Oh I see. So we should hold this PR and just add warning info to users at this moment right?

I think so

**Comment by ahengljh** [comment]
> > > As a start, maybe it would be better to just provide a deprecation warning for `lora_extra_vocab_size`
> > 
> > 
> > Oh I see. So we should hold this PR and just add warning info to users at this moment right?
> 
> I think so

I just made another one-line PR #23635 for adding the warning, and we may hold this PR until you believe it's time to deprecate this feature completely. What's your opinion on this?

**Comment by ahengljh** [comment]
> > > > As a start, maybe it would be better to just provide a deprecation warning for `lora_extra_vocab_size`
> > > 
> > > 
> > > Oh I see. So we should hold this PR and just add warning info to users at this moment right?
> > 
> > 
> > I think so
> 
> I just made another one-line PR #23635 for adding the warning, and we may hold this PR until you believe it's time to deprecate this feature completely. What's your opinion on this?

The #23635 for add warning message is merged. Now we may talk about how about changes in this PR? If we would deprecate this feature completely in v0.12.0, what else should be ready adding to this PR? So that I can make it and waiting for the time.

**Comment by zhuohan123** [comment]
@ahengljh Any blockers for us to go ahead, pass CI, and merge this PR? Happy to help on any blockers to merge the PR.

**Comment by jeejeelee** [comment]
> @ahengljh Any blockers for us to go ahead, pass CI, and merge this PR? Happy to help on any blockers to merge the PR.

Initially, the idea was to first add deprecation warnings,see https://github.com/vllm-project/vllm/pull/23635, then remove this code after some time, which might be more user-friendly. But if you want to remove this feature as soon as possible, I think that's also fine

**Comment by ahengljh** [comment]
> > @ahengljh Any blockers for us to go ahead, pass CI, and merge this PR? Happy to help on any blockers to merge the PR.
> 
> Initially, the idea was to first add deprecation warnings,see #23635, then remove this code after some time, which might be more user-friendly. But if you want to remove this feature as soon as possible, I think that's also fine

I suppose we were planning to proceed this PR in next release version so I hold it. If we would merge it soon, I’ll fix this PR asap.

**Comment by zhuohan123** [comment]
@ahengljh @jeejeelee Thanks for the context! I would suggest we just go ahead and merge it since we haven't received any objection in https://github.com/vllm-project/vllm/issues/23474 for two weeks and from the discussion there it should be a very rare feature from LoRA stats on HuggingFace.

Will leave the final timing decision to @aarnphm, @ywang96, and @jeejeelee

**Comment by zhuohan123** [comment]
@ahengljh We are ready to merge the code now! Could you help resolve the conflicts so we can merge? Thanks for the work!

**Comment by jeejeelee** [comment]
@ahengljh  If you are not available, I'll continue pushing forward with this PR tomorrow

**Comment by ahengljh** [comment]
> @ahengljh If you are not available, I'll continue pushing forward with this PR tomorrow

I will continue working on this, don’t worry

File: vllm/engine/arg_utils.py [hunk 1]
```diff
@@ -398,7 +398,6 @@ class EngineArgs:
     fully_sharded_loras: bool = LoRAConfig.fully_sharded_loras
     max_cpu_loras: Optional[int] = LoRAConfig.max_cpu_loras
     lora_dtype: Optional[Union[str, torch.dtype]] = LoRAConfig.lora_dtype
-    lora_extra_vocab_size: int = LoRAConfig.lora_extra_vocab_size
```
Comment (hmellor) [inline vllm/engine/arg_utils.py:453]: Can you also remove the field from `LoRAConfig`?

Comment (ahengljh) [reply vllm/engine/arg_utils.py:453]: Done, but I'm not quite sure about changes in vllm/lora/layers/logits_processor.py and vllm/lora/layers/vocal_parallel_embedding.py, please help review

**Comment by zhuohan123** [comment]
@ahengljh Just FYI #24078 is merged so we don't have tokenizer group in vLLM anymore.

**Comment by hmellor** [comment]
@jeejeelee could you give this a review to ensure LoRA will still behave as expected with the additional vocab feature removed?

File: vllm/lora/layers/vocal_parallel_embedding.py [hunk 2]
```diff
@@ -30,37 +30,13 @@ def create_lora_weights(
...
-                max_loras,
-                lora_config.lora_extra_vocab_size,
-                self.base_layer.embedding_dim,
-            ),
-            dtype=self.base_layer.weight.dtype,
-            device=self.base_layer.weight.device,
-        )
+        # No additional vocabulary support for LoRA
+        self.embeddings_slice = None
+        self.embeddings_weights = None
```
Comment (zhuohan123) [inline vllm/lora/layers/vocal_parallel_embedding.py]: Can we just remove these fields instead of leaving it as None?

File: vllm/lora/layers/logits_processor.py [hunk 3]
```diff
@@ -103,20 +103,13 @@ def create_lora_weights(
...
             dtype=lora_config.lora_dtype,
             device=self.device,
         )
-        self.embeddings_tensors = torch.full(
-            (max_loras, lora_config.lora_extra_vocab_size, self.hidden_size),
-            fill_value=float("-inf"),
-            dtype=self.dtype,
-            device=self.device,
-        )
+        # No additional vocabulary support for LoRA
```
Comment (zhuohan123) [inline vllm/lora/layers/logits_processor.py]: No need to leave comments like this here. For deprecated feature let's just make it disappear so later people reading the code does not get confused

File: tests/lora/test_layers.py [hunk 4]
```diff
@@ -566,8 +553,7 @@ def _pretest():
                       org_vocab_size:logits_processor.org_vocab_size +
                       embeddings_tensor_len] = embeddings_tensor
 
-        logits_processor.org_vocab_size = (vocab_size +
-                                           lora_config.lora_extra_vocab_size)
+        logits_processor.org_vocab_size = vocab_size
```
Comment (zhuohan123) [inline tests/lora/test_layers.py:556]: `org_vocab_size` Is this field only needed for lora additional vocab?

File: tests/lora/test_layers.py [hunk 5]
```diff
@@ -293,8 +293,7 @@ def create_random_embedding_layer():
                                    prompt_mapping,
                                    is_prefill=stage)
         punica_wrapper.update_metadata(lora_mapping, id_to_index, max_loras,
-                                       vocab_size,
-                                       lora_config.lora_extra_vocab_size)
+                                       vocab_size)
```
Comment (jeejeelee) [inline tests/lora/test_layers.py:296]: Have you tested these changes?

File: vllm/lora/layers/logits_processor.py (hunk 3 above)
Comment (jeejeelee) [reply vllm/lora/layers/logits_processor.py]: Reasonable, also please remove these comments from the model.

File: tests/lora/test_layers.py (hunk 5 above)
Comment (ahengljh) [reply tests/lora/test_layers.py:296]: Following the removal of LoRA additional vocabulary support, the following
   tests in tests/lora/test_layers.py are now failing:

  - test_embeddings_with_new_embeddings - Tests embedding expansion with
  additional vocabulary tokens
  - test_lm_head_logits_processor - Tests logits processing with extended
  vocabulary

  These tests specifically validate the removed functionality where LoRA
  adapters could introduce new tokens beyond the base model's vocabulary
  through embeddings_tensor.

I am not quite sure about these

Comment (jeejeelee) [reply tests/lora/test_layers.py:296]: Okay, I will check it out ASAP

**Comment by jeejeelee** [comment]
I have trigged the lora testing to verify these changes

**Comment by jeejeelee** [comment]
All LoRA tests are failing. Could you take a closer look into this? @ahengljh

**Comment by ahengljh** [comment]
> All LoRA tests are failing. Could you take a closer look into this? @ahengljh

Yes, I noticed that and will try to fix them

**Comment by jeejeelee** [comment]
@ahengljh  could we speed up a bit? Thank you.

**Comment by ahengljh** [comment]
> @ahengljh could we speed up a bit? Thank you.

I have been trying run the lora tests on my local machine but I found that even I ran them on the main branch, they still fail, which seems difficult for me to debug. I need to investigate and try more.
